core/lib/multivm/src/versions/vm_latest/types/internals/pubdata.rs

```diff
@@ -1,4 +1,10 @@
-use zksync_types::writes::{compress_state_diffs, StateDiffRecord};
+use zksync_mini_merkle_tree::MiniMerkleTree;
```

Comment:
Not really a fan of the `PubdataInput` dataflow. Optimally, I'd imagine there would be a trait mapping `PubdataInput` to `Vec<u8>` published pubdata (for bootloader memory and external consumption in `FinishedL1Batch`; maybe 2 versions if there is no well-defined transform between both; or maybe it makes sense to change `FinishedL1Batch.pubdata_input` type to `Option<PubdataInput>` to only keep the bootloader memory version). This trait would be used by the VM similarly to storage (probably with dynamic dispatch; it doesn't look performance-critical). This would allow to not thread `PubdataParams` everywhere and to not include additional deps for `multivm`. OTOH, I suspect this would require non-trivial amount of refactoring.

Reply:
I agree with you and IMO you described the end goal accurately. It doesn't seem like something is blocking it, I will try to implement

Reply:
PTAL, I have updated PR. Now it uses `Rc<dyn PubdataBuilder>` inside VM (`Rc` is used instead of `Box` only  because I need `Clone` for shadow VM), and converts `PubdataParams` to `PubdataBuilder` in all places where VM is inited